Concept docs for "Numbers" (closes #600) (#602)
exercism/cpp: concepts/numbers/introduction.md

# Introduction
The built-in number types in C++ can be divided into integers and floating points.
Integers are whole numbers like `0`, `691`, or `-2`.
Examples of floating point numbers are `6.02214076`, `0.1`, or `-1.616`.

> ## Integers

Reviewer: Students coming from some other languages might expect `int`s to hold arbitrarily large integers. Suggestion: add a note about bounds.

Author: I added it as a section of the about.md, as it is not necessary to solve the exercise.
I plan to to a special concept on unsigned, limits, bit-width etc. So there they can experience it.

When you assign a value to an `int` variable, you can do so directly with a literal.
There are different integer literals based on the base of the representation.
Decimal integer literals are the most common and use the numbers `0` to `9`.
By adding a special prefix it is possible to use other bases.
The example below shows the number `9421` in its four representations and prefixes.

> 

Reviewer: With my Beginner hat on, I have now finished the above paragraph but am not much wiser. I think examples first would be better: then it would be clearer what the words (literal, decimal literal, prefixes, bases, representations) refer to.

```cpp
int decimal{9241};              // base 10: 0-9
int hexadecimal{0x24CD};        // base 16: 0-9 and A-F
int binary{0b10010000011001};   // base  2: 0-1

> int octal{022031};              // base  8: 0-7

Reviewer 2: Yes, it is mentioned in the basic concept: https://github.com/exercism/cpp/pull/594

                                // Leading with a 0 not the letter o.
```
You can use an apostrophe to separate digits for easier readability.
`9'241` is the same as `0b0100'100'0001'1001` or `92'4'1`.

> ## Floating-Point Numbers

Reviewer: Suggestion: add a note along the lines of

> C++'s floats are IEEE 754 floats, just like in most other programming languages.
> These have finite precision (namely roughly such and such significant digits) and sometimes display weird behavior, such as `0.1 + 0.2 != 0.3` (link).

Author: I have some information in the about.md.
It's not so clean cut in C++. If you want a deep-dive:
"Unfortunately, C++ standard guarantees almost nothing about the built-in floating point types."
https://kkimdev.github.io/posts/2018/06/15/IEEE-754-Floating-Point-Type-in-C++.html

The floating-point literals come in two flavors. 

> Next to the intuitive `0.0024` it is possible to use its scientific notation `2.4e-3`.

Reviewer: ```suggestion
In addition to the intuitive `0.0024` it is possible to use its scientific notation `2.4e-3`.
```

The most common floating-point type is `double`.

## Arithmetic

C++ supports `+`, `-`, `*`, `/`, `(` and `)` and `%` to form expressions.
The result from the operation between two integers is also an integer.
`5 / 2` will return `2`.
When one of the involved types is a floating-point type, the result will also be of a floating-point.
`5.0 / 2` and `5 / 2.0` will return `2.5`.
`%` is the modulus operator and will return the remainder of an integer division: `5%3` is `2`.

## The auto keyword

It is possible to let the compiler figure out the correct type of a variable at the point of declaration with the `auto` keyword.
The practice can help during refactoring and shines with very long type names, that will come in later concepts.

```cpp

> auto feist{1234};             // feist is of type int, as 27 is an integer literal

Reviewer 3: ```suggestion
auto feist{1234};             // feist is of type int, as 1234 is an integer literal
```

# Introduction
The built-in number types in C++ can be divided into integers and floating points.
Integers are whole numbers like `0`, `691`, or `-2`.

> Examples of floating point numbers are `6.02214076`, `0.1`, or `-1.616`.

Reviewer 2: ```suggestion
Floating point numbers are decimal numbers like `6.02214076`, `0.1`, or `-1.616`.
```

Feels a bit more consistent to write it like this

int a_apaec{0x24CD};              // base 16: 0-9 and A-F
int m_gargan{0b10010000011001};   // base  2: 0-1
int b_reilly{022031};             // base  8: 0-7
                                  // Leading with a 0 not the letter o.
```
When you assign a value to an `int` variable, you can do so directly with a literal.
A literal is a hard-coded number like `9241`.
There are different integer literals for several bases of the representation.
Decimal integer literals are the most common and use the digits `0` to `9`.
By adding a special prefix, like `0x`, it is possible to use other bases.
The example above shows the number `9421` in its four representations and prefixes.
All variables are initialized to the same value.

For more details on the different representation systems, take a look at [a small tutorial][cpp_numerical_bases].

You can use an apostrophe to separate digits for easier readability.
`9'241` is the same as `0b0100'100'0001'1001` or `92'4'1`.

## Floating-Point Numbers

The floating-point literals come in two flavors. 
In addition to the intuitive `0.0024` it is possible to use its scientific notation `2.4e-3`.
The most common floating-point type is `double`.

> ## Arithmetic

Reviewer 2: If C++ follows the PEMDAS standard I think that could be mentioned: https://en.wikipedia.org/wiki/Order_of_operations (although requires the language to have expoentials)

Author: For C++ irt would be PMDAS. No E :(

C++ supports `+`, `-`, `*`, `/`, `(` and `)` and `%` to form expressions.
The result from the operation between two integers is also an integer.
`5 / 2` will return `2`.
When one of the involved types is a floating-point type, the result will also be of a floating-point.
`5.0 / 2` and `5 / 2.0` will return `2.5`.

> `%` is the modulus operator and will return the remainder of an integer division: `5%3` is `2`.

Reviewer 4: Nitpick: technically, the C++ `%` operator is the remainder operator, not modulus. (see languages like Julia which have both [remainder](https://docs.julialang.org/en/v1/base/math/#Base.rem) and [modulus](https://docs.julialang.org/en/v1/base/math/#Base.mod) functions.)

Reviewer: Aye. The behavior of `%` matches with that of `/` via `a = q n + r` if `q = a / n` and `r = a % n`.

The difference between the two conventions:

```text
                       /   %                              mod
 7 `quotRem`  3  ==  ( 2,  1)     7 `divMod`  3  ==  ( 2,  1)
 7 `quotRem` -3  ==  (-2,  1)     7 `divMod` -3  ==  (-3, -2)
-7 `quotRem`  3  ==  (-2, -1)    -7 `divMod`  3  ==  (-3,  2)
-7 `quotRem` -3  ==  ( 2, -1)    -7 `divMod` -3  ==  ( 2, -1)

```

> ## The auto keyword

Reviewer 5: Are we sure we want to introduce `auto` with numbers? Or it's already been introduced and we just talk about it cause it's something good to know?

It is possible to let the compiler figure out the correct type of a variable at the point of declaration with the `auto` keyword.
The practice can help during refactoring and shines with very long type names, that will come in later concepts.

```cpp
auto feist{1234};             // feist is of type int, as 1234 is an integer literal
auto red_cross{0xED1B2E};     // also an int
auto pythargoras{1.41421356}; // pythargoras is of type double

> auto two_times(int number) {

Reviewer 4: I think using `auto` to deduce function return types is a different topic than using it to deduce variable types, and should perhaps be kept for later when functions are more specifically mentioned.

Reviewer 5: +1 on this, it kinda makes sense to introduce `auto` in this exercise but using it in functions it's quite a stretch.
